File: Recursion/Day10/wordBreak.cpp
// https://www.codingninjas.com/codestudio/problems/983635
// Time -
// Space -

#include <bits/stdc++.h>

void solve(int index, int n, string &s, map<string, bool> m, vector<string> &ans, string combi)
{

    if (index == n)
    {
        ans.push_back(combi);
    }

    for (int i = index; i < n; i++)
    {
        string sub = s.substr(index, i - index + 1);
        if (m[sub])
        {
            solve(i + 1, n, s, m, ans, combi + sub + " ");
        }
    }
}

vector<string> wordBreak(string &s, vector<string> &dictionary)
{
    // Write your code here
    vector<string> ans;
    map<string, bool> m;

    for (int i = 0; i < dictionary.size(); i++)
    {
        m[dictionary[i]] = true;
    }
    string ep = "";
    solve(0, s.size(), s, m, ans, ep);

    return ans;
}

File: DP/Day26/wordBreak.js
/**
 * @param {string} A
 * @param {string[]} B
 * @returns {number}
 * https://practice.geeksforgeeks.org/problems/word-break1352/1
 * Time - O(M*N)
 * Space - O(M) recursive space
 */

class Solution {
  wordBreak(A, B) {
    //code here
    return this.solve(0, A, B, A.length, B.length);
  }

  solve(index, A, B, m, n) {
    if (index >= m) return 1;

    let ans = 0;
    for (let i = 0; i < n; i++) {
      const two = B[i];
      const one = A.substring(index, index + two.length);

      if (one == two) {
        ans = ans || this.solve(index + two.length, A, B, m, n);
      }
    }

    return ans;
  }
}
